### PR TITLE
feat(access): allow more flexible access checks

### PR DESCRIPTION
**Pull request by oyo**
## Description

Enable access checks with arbitrary check functions.

## Why

Currently the frontend role check is only based on role name assuming the roles names are unique across clients. This assumption can be incorrect so we implemented a more flexible access check which allows to check for client id and role name. To be future-proof the mechanism now allows arbitrary checks to be executed for decision whether access is granted or not.

In the process some redundant code has been identified and removed.

## Issue

#657 

## Checklist

- [x] I have followed the [contributing guidelines](https://github.com/eclipse-tractusx/portal-assets/blob/main/docs/developer/Technical%20Documentation/Dev%20Process/How%20to%20contribute.md#commit-and-pr-guidelines)
- [x] I have performed a self-review of my own code
- [x] I have successfully tested my changes locally
- [x] I have added tests that prove my changes work
- [x] I have checked that new and existing tests pass locally with my changes
- [x] I have commented my code, particularly in hard-to-understand areas

### CODE REVIEW

File: src/components/pages/AppDetail/components/AppDetailHeader/index.tsx [hunk 1]
```diff
@@ -92,11 +86,12 @@ export default function AppDetailHeader({ item }: AppDetailHeaderProps) {
...
         btnColor = {
-          color:
-            user.roles.indexOf(Roles.SUBSCRIBE_APPS) !== -1 &&
-            user.roles.indexOf(Roles.SUBSCRIBE_SERVICE) !== -1
-              ? 'primary'
-              : 'secondary',
+          color: userHasPortalRole([
+            ROLES.SUBSCRIBE_APPS,
+            ROLES.SUBSCRIBE_SERVICE,
+          ])
```
Comment (oyo) [inline src/components/pages/AppDetail/components/AppDetailHeader/index.tsx:92]: @jjeroch Not sure what this is good for but I think it was implemented incorrectly setting the button color only if the user has both roles. Now the check is from AND to OR setting the color if the user has at least one of the two roles.

**Review by evegufy — COMMENTED** [review]
Looks good in general but as the impact of this change could be quite huge if there was some inconsistency in which client was validated updated to now: did you check the following?: 
"
The following resource needs to get checked when the user access the respective page:

| URL                   | Ressource           |
|-----------------------|---------------------|
| /registration         | Cl1-CX-Registration |
| /registration/*       | Cl1-CX-Registration |
| /?overlay=consent_osp | Cl1-CX-Registration |
| /semantichub          | Cl3-CX-Semantic     |
| /admin-credential            | Cl24-CX-SSI-Credential-Issuer       |
| /usecase-participation            | Cl2-CX-Portal        |
| /certificate-credential            | Cl2-CX-Portal        |
| all others            | Cl2-CX-Portal       |

"

**Comment by oyo** [comment]
> Looks good in general but as the impact of this change could be quite huge if there was some inconsistency in which client was validated updated to now: 

The logic now intersects the resourceAccess attribute from the keycloak user token with a same object securing any resource. Access is granted if the intersection is not empty which means that at least one required client+role combination has been found in the user token. Tests for this logic are here:

https://github.com/eclipse-tractusx/portal-frontend/blob/feat/657/allow-clientid-for-rolecheck/src/utils/dataUtils.test.ts#L169-L176

> did you check the following?: " The following resource needs to get checked when the user access the respective page:
> 
> URL	Ressource
> /registration/*	Cl1-CX-Registration

Those URLs are from the registration app and the check must be implemented there.

> /registration	Cl1-CX-Registration

https://github.com/eclipse-tractusx/portal-frontend/blob/feat/657/allow-clientid-for-rolecheck/src/types/Config.tsx#L111-L112

> /?overlay=consent_osp	Cl1-CX-Registration

Looks like this overlay is not implemented any more. 
https://portal.dev.demo.catena-x.net/?overlay=consent_osp
There is a page with this name - but it's broken
https://portal.dev.demo.catena-x.net/consent_osp
@jjeroch has this logic been changed or is it broken for other reasons?

> /semantichub	Cl3-CX-Semantic

https://github.com/eclipse-tractusx/portal-frontend/blob/feat/657/allow-clientid-for-rolecheck/src/types/Config.tsx#L169-L175

> /admin-credential	Cl24-CX-SSI-Credential-Issuer

https://github.com/eclipse-tractusx/portal-frontend/blob/feat/657/allow-clientid-for-rolecheck/src/types/Config.tsx#L567-L568

> /usecase-participation	Cl2-CX-Portal
> /certificate-credential	Cl2-CX-Portal
> all others	Cl2-CX-Portal

all other pages are using `userHasPortalRole` checking for that client id if correctly injected:
https://github.com/eclipse-tractusx/portal-frontend/blob/feat/657/allow-clientid-for-rolecheck/src/services/AccessService.tsx#L97-L98

File: src/components/pages/AdminCredential/AdminCredentialElements.tsx [hunk 2]
```diff
@@ -303,7 +303,7 @@ export default function AdminCredentialElements() {
       flex: 1,
       renderCell: ({ row }: { row: CredentialData }) => (
         <>
-          {UserService.hasRole(ROLES.REVOKE_CREDENTIALS_ISSUER) &&
+          {userHasPortalRole(ROLES.REVOKE_CREDENTIALS_ISSUER) &&
```
Comment (jjeroch) [inline src/components/pages/AdminCredential/AdminCredentialElements.tsx:306]: @oyo this role is the role of the issuer component
image

File: src/types/Config.tsx [hunk 3]
```diff
@@ -110,28 +120,24 @@ export const ALL_PAGES: IPage[] = [
...
     element: <Redirect path="documentation" tab={'documentation'} />,
   },
-  {
-    name: PAGES.STORYBOOK,
-    element: <Redirect path="_storybook" tab={'storybook'} />,
-  },
   {
     name: PAGES.MARKETPLACE,
-    role: ROLES.APPSTORE_VIEW,
+    allowTo: () => userHasPortalRole(ROLES.APPSTORE_VIEW),
```
Comment (jjeroch) [inline src/types/Config.tsx:125]: @oyo could you please recheck what APPSTORE_VIEW is? I dont think we have such an role. The actual permission name is `view_apps`

File: src/types/Config.tsx [hunk 4]
```diff
@@ -110,28 +120,24 @@ export const ALL_PAGES: IPage[] = [
...
   {
     name: PAGES.APP_MARKETPLACE,
-    role: ROLES.APPSTORE_VIEW,
+    allowTo: () => userHasPortalRole(ROLES.APPSTORE_VIEW),
     element: <AppMarketplace />,
   },
   {
     name: PAGES.SERVICE_MARKETPLACE,
-    role: ROLES.APPSTORE_VIEW_SERVICES,
+    allowTo: () => userHasPortalRole(ROLES.APPSTORE_VIEW_SERVICES),
```
Comment (jjeroch) [inline src/types/Config.tsx:135]: @oyo  could you please recheck what APPSTORE_VIEW_SERVICES is? I dont think we have such an role. The actual permission name is `view_service_marketplace`

File: src/types/Config.tsx [hunk 5]
```diff
@@ -592,117 +608,117 @@ export const ALL_PAGES: IPage[] = [
...
-    role: ROLES.MODIFY_USER_ACCOUNT,
+    allowTo: () => userHasPortalRole(ROLES.MODIFY_USER_ACCOUNT),
   },
-  { name: OVERLAYS.APP, role: ROLES.APPSTORE_VIEW },
+  { name: OVERLAYS.APP, allowTo: () => userHasPortalRole(ROLES.APPSTORE_VIEW) },
   { name: OVERLAYS.NEWS },
   {
     name: OVERLAYS.PARTNER,
-    role: ROLES.PARTNER_NETWORK_VIEW,
+    allowTo: () => userHasPortalRole(ROLES.PARTNER_NETWORK_VIEW),
```
Comment (jjeroch) [inline src/types/Config.tsx:649]: @oyo  here we need the BPDM role inside the user account.
image

File: src/types/Config.tsx [hunk 6]
```diff
@@ -181,87 +187,93 @@ export const ALL_PAGES: IPage[] = [
...
   {
     name: PAGES.ORGANIZATION,
-    role: ROLES.MY_ORGANIZATION_VIEW,
+    allowTo: () => userHasPortalRole(ROLES.MY_ORGANIZATION_VIEW),
     element: <Organization />,
   },
   {
     name: PAGES.PARTNER_NETWORK,
-    role: ROLES.PARTNER_NETWORK_VIEW,
+    allowTo: () => userHasPortalRole(ROLES.PARTNER_NETWORK_VIEW),
```
Comment (jjeroch) [inline src/types/Config.tsx:210]: @oyo  here we need the BPDM role inside the user account.
image

File: src/types/Config.tsx [hunk 7]
```diff
@@ -533,38 +549,38 @@ export const ALL_PAGES: IPage[] = [
...
   {
     name: PAGES.COMPANY_CERTIFICATE,
-    role: ROLES.COMPANY_CERTIFICATE_VIEW,
+    allowTo: () => userHasPortalRole(ROLES.COMPANY_CERTIFICATE_VIEW),
     element: <CompanyCertificates />,
   },
   {
     name: PAGES.COMPANY_WALLET,
-    role: ROLES.CREDENTIAL_REQUESTS,
+    allowTo: () => userHasPortalRole(ROLES.CREDENTIAL_REQUESTS),
```
Comment (jjeroch) [inline src/types/Config.tsx:583]: @oyo update needed - access is
image

**Review by jjeroch — CHANGES_REQUESTED** [review]

File: src/components/pages/AppDetail/components/AppDetailHeader/index.tsx (hunk 1 above)
Comment (oyo) [reply src/components/pages/AppDetail/components/AppDetailHeader/index.tsx:92]: @jjeroch yes, there is the same check further below if the button is selectable or not

File: src/components/pages/CompanyCertificates/CompanyCertificateCard.tsx [hunk 8]
```diff
@@ -171,7 +171,7 @@ export default function CompanyCertificateCard({
           borderRadius: '74px',
         }}
         onClick={handleDelete}
-        disabled={!UserService.hasRole(ROLES.SUBSCRIBE_SERVICE_MARKETPLACE)}
+        disabled={!userHasPortalRole(ROLES.SUBSCRIBE_SERVICE)}
```
Comment (jjeroch) [inline src/components/pages/CompanyCertificates/CompanyCertificateCard.tsx:174]: that looks awkward. The user should get the "Subscribe Service" Role to access the Company Certificates? @oyo  lets quickly chat here later today

**Review by jjeroch — CHANGES_REQUESTED** [review]

**Comment by jjeroch** [comment]
@oyo beside the comment added to the code; one more thing

- Wallet UI - the revocation button should also need a `{userHasSsiCredentialRole` could you please check this once
- Partner Network - for the Partner Network UI the user needs to have BPDM roles assigned; do we have this validation somewhere included? I couldnt see it

File: src/components/pages/CompanyCertificates/CompanyCertificateCard.tsx (hunk 8 above)
Comment (oyo) [reply src/components/pages/CompanyCertificates/CompanyCertificateCard.tsx:174]: @jjeroch changed this to DELETE_CERTIFICATES

**Comment by oyo** [comment]
> * Wallet UI - the revocation button should also need a `{userHasSsiCredentialRole` could you please check this once

added `userHasSsiCredentialRole(ROLES.REVOKE_CREDENTIALS_ISSUER)` to the check

> * Partner Network - for the Partner Network UI the user needs to have BPDM roles assigned; do we have this validation somewhere included? I couldnt see it

it is implemented with the check `allowTo: () => userHasBpdmRole(ROLES.READ_PARTNER),`
https://github.com/eclipse-tractusx/portal-frontend/pull/873/files#diff-f207e98afeed11e7e08f809e36fa97a268e311ceeeebb779c00d05f3bf99ba6aR210

**Review by jjeroch — APPROVED** [review]

**Review by evegufy — CHANGES_REQUESTED** [review]
@oyo could you please resolve the conflicts? I'd approve it then.

**Comment by oyo** [comment]
> @oyo could you please resolve the conflicts? I'd approve it then.

@evegufy conflicts resolved and PR updated

**Review by evegufy — APPROVED** [review]

**Pull request merged**